libr/core/cmd_debug.c

```diff
@@ -792,6 +796,224 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
 	r_list_free (list);
 }
 
+static void update_main_arena(RCore *core, ut64 m_arena, RHeap_MallocState *main_arena) {
+	r_core_read_at (core, m_arena, (ut8 *)main_arena, sizeof (RHeap_MallocState));
+}
+
+#define PRINTF_ARENA(color, fmt , ...) r_cons_printf (color fmt Color_RESET, __VA_ARGS__)
+#define PRINTF_GREEN_ARENA(fmt, ...) PRINTF_ARENA (Color_GREEN, fmt, __VA_ARGS__)
+#define PRINTF_BLUE_ARENA(fmt, ...) PRINTF_ARENA (Color_BLUE, fmt, __VA_ARGS__)
+
+#define PRINT_ARENA(color, msg) r_cons_print (color msg Color_RESET)
+#define PRINT_GREEN_ARENA(msg) PRINT_ARENA (Color_GREEN, msg)
+#define PRINT_BLUE_ARENA(msg) PRINT_ARENA (Color_BLUE, msg)
+static void print_main_arena(ut64 m_arena, RHeap_MallocState *main_arena) {
+	int i, offset;
+	PRINT_GREEN_ARENA ("main_arena @ ");
+	PRINTF_BLUE_ARENA ("0x%"PFMT64x"\n\n", (ut64)(size_t)m_arena);
+	PRINT_GREEN_ARENA ("struct malloc_state main_arena {\n");
+	PRINT_GREEN_ARENA ("\tmutex = ");
+	PRINTF_BLUE_ARENA (" 0x%x\n", (int)main_arena->mutex);
+	PRINT_GREEN_ARENA ("\tflags = ");
+	PRINTF_BLUE_ARENA (" 0x%x\n", (int)main_arena->flags);
+	PRINT_GREEN_ARENA ("\tfastbinsY = {");
+
+	for (i = 0; i < NFASTBINS; i++) {
+		PRINTF_BLUE_ARENA ("0x%"PFMT64x, (ut64)(size_t)main_arena->fastbinsY[i]);
+		if (i < 9) {
+			PRINT_GREEN_ARENA (",");
+		}
+	}
+	PRINT_GREEN_ARENA ("}\n");
+	PRINT_GREEN_ARENA ("\ttop = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->top);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("\tlast_remainder = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->last_remainder);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("\tbins {");
+
+	offset = (size_t)&main_arena->last_remainder - (size_t)&main_arena->mutex + sizeof (size_t);
+	for (i = 0; i < NBINS * 2 - 2; i += 2) {
+		if (i % 2 == 0) r_cons_print ("\n\t");
+		if (!main_arena->bins[i]) {
+			PRINT_BLUE_ARENA ("0x0 ");
+			PRINT_GREEN_ARENA ("<repeats 254 times>");
+			break;
+		} else {
+			ut64 tmp = (size_t)m_arena + (size_t)offset + sizeof (size_t) * i - sizeof (size_t) * 2;
+			PRINTF_GREEN_ARENA (" 0x%"PFMT64x"->fd = ", tmp);
+			PRINTF_BLUE_ARENA ("0x%"PFMT64x, (ut64)(size_t)main_arena->bins[i]);
+			PRINT_GREEN_ARENA (", ");
+			PRINTF_GREEN_ARENA (" \t0x%"PFMT64x"->bk = ", tmp);
+			PRINTF_BLUE_ARENA ("0x%"PFMT64x, (ut64)(size_t)main_arena->bins[i + 1]);
+			PRINT_GREEN_ARENA (", ");
+		}
+	}
+	PRINT_GREEN_ARENA ("\n\t}\t\n");
+	PRINT_GREEN_ARENA ("\tbinmap = {");
+
+	for(i = 0; i < BINMAPSIZE; i++) {
+		PRINTF_BLUE_ARENA ("0x%x", (int)main_arena->binmap[i]);
+		if (i < 3) {
+			PRINT_GREEN_ARENA (",");
+		}
+	}
+	PRINT_GREEN_ARENA ("}\n");
```

Comment:
use shorter macro names. like single uppercase letter. G instead of PRINT_GREEN_ARENA and BF for PRINTF_BLUE_ARENA. for example